src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs

```diff
@@ -1997,23 +1998,27 @@ internal void SetDependencyNodeTargetIdentifier(ProjectId projectId, string targ
             }
         }
 
-        internal void RefreshProjectExistsUIContextForLanguage(string language)
+        /// <summary>
+        /// Must only be called when caller is holding <see cref="_gate"/>
+        /// </summary>
+        internal async Task RefreshProjectExistsUIContextForLanguage_NoLockAsync(string language, CancellationToken cancellationToken)
         {
-            // We must assert the call is on the foreground as setting UIContext.IsActive would otherwise do a COM RPC.
+            Contract.ThrowIfTrue(!Monitor.IsEntered(_gate) && Monitor.TryEnter(_gate), "Caller was not already holding the _gate");
```

Comment:
I'm not sure what's going on here. If this is called without the gate being held, then we'll call TryEnter and if that succeeds throw, but at that point the gate is held permanently?

Reply:
Also if somebody else is holding the gate then this will evaluate to `!false && false` and march right along.